common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java

```diff
@@ -825,45 +881,79 @@ void resetChunkTracker() {
+        writeChunkTracker(mapIndex);
+        indexFile.updatePos(8);
+        this.lastChunkOffset = chunkOffset;
+        indexMetaUpdateFailed = false;
+      } catch (IOException ioe) {
+        logger.warn("{} shuffleId {} reduceId {} update to index/meta failed", appShuffleId.appId,
+          appShuffleId.shuffleId, reduceId);
+        indexMetaUpdateFailed = true;
+        // Any exception here is propagated to the caller and the caller can decide whether to
+        // abort or not.
         throw ioe;
       }
-      this.lastChunkOffset = chunkOffset;
     }
 
     private void writeChunkTracker(int mapIndex) throws IOException {
       if (mapIndex == -1) {
         return;
       }
       chunkTracker.add(mapIndex);
-      long metaStartPos = metaFile.getFilePointer();
       try {
         logger.trace("{} shuffleId {} reduceId {} mapIndex {} write chunk to meta file",
           appShuffleId.appId, appShuffleId.shuffleId, reduceId, mapIndex);
-        chunkTracker.serialize(metaFile);
+        if (indexMetaUpdateFailed) {
+          metaFile.getChannel().position(metaFile.getPos());
+        }
+        int chunkTrackerSize = chunkTracker.serializedSizeInBytes();
+        chunkTracker.serialize(metaFile.getDos());
+        metaFile.updatePos(chunkTrackerSize);
       } catch (IOException ioe) {
-        logger.warn("{} shuffleId {} reduceId {} mapIndex {} reset position of meta file to {}",
-          appShuffleId.appId, appShuffleId.shuffleId, reduceId, mapIndex, metaStartPos);
-        metaFile.seek(metaStartPos);
+        logger.warn("{} shuffleId {} reduceId {} mapIndex {} update to meta file failed",
+          appShuffleId.appId, appShuffleId.shuffleId, reduceId, mapIndex);
+        // Any exceptions while writing to meta file is propagated to updateChunkInfo so that
+        // the index file position can be reset as well.
         throw ioe;
       }
     }
 
+    private void incrementIOExceptions() {
+      numIOExceptions++;
+    }
+
+    private boolean shouldAbort(int ioExceptionsThresholdDuringMerge) {
+      return numIOExceptions > ioExceptionsThresholdDuringMerge;
+    }
+
+    private void finalizePartition() throws IOException {
+      if (dataFilePos != lastChunkOffset) {
+        try {
+          updateChunkInfo(dataFilePos, lastMergedMapIndex);
+        } catch (IOException ioe) {
+          // Any exceptions here while updating the meta files can be ignored. If the files
+          // aren't successfully updated they will be truncated.
+        }
+      }
+      // Get rid of any partial block data at the end of the file. This could either
+      // be due to failure, or a request still being processed when the shuffle
+      // merge gets finalized, or any exceptions while updating index/meta files.
+      dataChannel.truncate(lastChunkOffset);
+      indexFile.getChannel().truncate(indexFile.getPos());
+      metaFile.getChannel().truncate(metaFile.getPos());
```

Comment:
We are not ignoring the exceptions for truncate. If any of these truncate fail, the finalize of this shuffle partition fails and the shuffle partition is considered not to be merged. So clients will fetch original un-merged data.

If we want, we can ignore exceptions during truncation of data file. However, I didn't do that because the existing code doesn't ignore that.

Reply:
I don't think we can ignore any IOExceptions during truncation of index/meta file right now. This is because how we serve the number of chunks from index file and we expect the number of chunks from index file and meta file to be consistent.

Reply:
Had offline discussion with @otterc ... ignoring truncate exceptions here will require more changes to how we compute and manage chunks, keep appShufflePartitionInfo around, etc ... 
Given this, I am fine with handling the exception and not ignoring it.
Thanks for details @otterc